mh/state.py

```diff
@@ -589,9 +792,518 @@ def match_city(city, fields):
+                get_ip(remote_config["IP"]), int(remote_config["Port"]))
+        else:
+            config = get_config("FMP")
+            get_instance().setup_server(server_id, "", 0, '0.0.0.0', config["Port"])
+        self.refresh_period = refresh_period
+        self.handlers = {}
+        self.central = server_id==0
+        self.server_id = server_id
+        Logger.__init__(self)
+        self.set_logger(create_logger("Cache", level=logging.DEBUG if debug_mode else logging.INFO,
+            log_to_file="cache.log" if log_to_file else "",
+            log_to_console=log_to_console,
+            log_to_window=log_to_window))
+        central_config = get_central_config()
+        self.central_connection = (central_config["CentralIP"], central_config["CentralCrossconnectPort"])
+        self.sel = selectors.DefaultSelector()
+
+    def update_player_record(self, session):
+        get_instance().update_capcom_id(session)
+
+    def update_players(self):
+        players = []
+        for server_id, server in self.servers.items():
+            if server_id != self.server_id:
+                players = players + server.get_all_players()
+        new_players = {}
+        for p in players:
+            new_players[p.capcom_id] = p
+        self.players = new_players
+        if self.server_id == 0:
+            get_instance().update_players()
+
+    def get_server_list(self):
+        servers = []
+        maxkey = 0
+        for i in self.servers.keys():
+            if i > maxkey:
+                maxkey = i
+        for i in range(maxkey):
+            if i+1 in self.servers:
+                servers.append(self.servers[i+1])
+            else:
+                servers.append(None)
+        return servers
+
+    def send_session_info(self, server_id, session):
+        self.outbound_sessions.append(
+            (server_id, json.dumps(session.serialize()))
+        )
+
+    def new_session(self, session):
+        get_instance().register_pat_ticket(session)
+
+    def get_handler(self, server_id):
+        return self.handlers[server_id]
+
+    def register_handler(self, server_id, handler):
+        self.handlers[server_id] = handler
+
+    def maintain_connection(self):
+        if self.central:
+            # CENTRAL SERVER CONNECTION TO REMOTE
+            while not get_instance().initialized:
+                pass
+            get_instance().cache = self
+            refresh_timer = Timer()
+            lsock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            try:
+                lsock.bind(self.central_connection)
+                lsock.listen(0)
+            except:
+                self.close()
+                raise
+            self.info("Listening for remote Servers on {}".format(self.central_connection))
+            lsock.setblocking(False)
+            self.sel.register(lsock, selectors.EVENT_READ, data=None)
```

Comment:
AFAICT, the socket is using TCP in plain without a security layer such as TLS, which is bad.

BTW, the function is way too big for what it doesn't and should be split in shorter methods and documented.

Reply:
This is now fixed, TLS has been integrated and the function has been split up